### PR TITLE
AP-329 Extract Kafka messages into fixed columns and generate MESSAGE_TIMESTAMP and PKs

### PR DESCRIPTION
**Pull request by koszti**
This PR fits to the pipelinewise YAML created in the following format:

```
---
id: "kafka_some_topic"
name: "Kafka - Some topic"
type: "tap-kafka"
owner: "somebody@transferwise.com"
sync_period: "*/30 * * * *"
db_conn:
  group_id: "analyticsDb"
  bootstrap_servers: "kafka-1.foo.com:9092,kafka-2.foo.com:9092"
  topic: "my.topic"

  # OPTIONAL: Extract primary key(s) into extra column(s) from the original Kafka message
  primary_keys:
    id: "$.properties.basicTransferMetadataDto.transferId"

# Defines target tables
schemas:
  - source_schema: "kafka"
    target_schema: "kafka"
    target_schema_select_permissions:
      - grp_stats
    tables:
      - table_name: "my_topic"
```

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -38,29 +38,39 @@ or
...
 
 ```
 {
   "group_id": "1",
   "bootstrap_servers": "foo.com,bar.com",
   "topic": "messages",
-  "reject_topic": "reject_topic",
-  "schema": "<json schema>"
+  "message_timestamp": "$.jsonpath.to.timestamp",
+  "primary_keys": {
```
Comment (ivan-transferwise) [inline README.md]: primary_key**s**? How do we specify multiple primary keys?

File: setup.py [hunk 2]
```diff
@@ -14,11 +14,15 @@
       install_requires=[
           'kafka-python',
           'singer-python==5.2.0',
-          'requests==2.20.0',
```
Comment (ivan-transferwise) [inline setup.py]: we don't need _requests_ anymore?

File: README.md (hunk 1 above)
Comment (louis-pie) [reply README.md]: does this mean composite primary key ?

File: tap_kafka/common.py [hunk 3]
```diff
@@ -1,19 +1,33 @@
...
+        "type": "object",
+        "properties": {
+            "message_timestamp": {"type": ["string", "null"], "format": "date-time"},
+            "message": {"type": ["object", "array", "string", "null"]}
+        }
+    }
+
+    # Primary keys are dynamic schema items
+    for pk in primary_keys:
+        schema["properties"][pk] = {"type": ["string", "null"]}
```
Comment (ivan-transferwise) [inline tap_kafka/common.py]: When primary_key could be null?

File: tests/test_tap_kafka.py [hunk 4]
```diff
@@ -0,0 +1,118 @@
+import pytest
```
Comment (ivan-transferwise) [inline tests/test_tap_kafka.py]: I think would be also really nice to check that tap produces all the messages as expected for some fake config :)

File: tap_kafka/common.py (hunk 3 above)
Comment (louis-pie) [reply tap_kafka/common.py]: Primary Key has become optional

File: README.md (hunk 1 above)
Comment (koszti) [reply README.md]: yes, adding 0, 1, or more columns are all supported. Singer spec supports composite primary keys and most of the tap support it anyway.

To specify multiple columns as composite PK:
  primary_keys:
    id: "$.properties.basicTransferMetadataDto.transferId"
    version: "$.properties.basicTransferMetadataDto.transferVersion"

File: setup.py (hunk 2 above)
Comment (koszti) [reply setup.py]: I hope so, everything that's required embedded in python-kafka

File: tap_kafka/common.py (hunk 3 above)
Comment (koszti) [reply tap_kafka/common.py]: exactly, thanks @louis-pie

File: tests/test_tap_kafka.py (hunk 4 above)
Comment (koszti) [reply tests/test_tap_kafka.py]: makes sense, I'm going to update the PR

File: README.md (hunk 1 above)
Comment (louis-pie) [reply README.md]: In other words, it is Primary JSON Keys ?

File: tests/test_tap_kafka.py (hunk 4 above)
Comment (koszti) [reply tests/test_tap_kafka.py]: addressed and added tests by https://github.com/transferwise/pipelinewise-tap-kafka/pull/5/commits/c19873e67d3fff22f3b4a11b435bc53609ba9064

File: README.md (hunk 1 above)
Comment (koszti) [reply README.md]: this means that two columns `id` and `version` will be added to the target table, the values will be extracted from the JSON message by the jsonpaths (`$.properties...`) and the two columns will be used as primary keys in the target table in target-snowflake|redshift|postgres

**Comment by koszti** [comment]
Modified tap-kafka to extract `MESSAGE_TIMESTAMP` values automatically from the kafka record metadata object in every message.

**Pull request merged**